User/pagoe/post build update (#698)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/concepts/userdatafolder.md

---
description: Learn about how to manage the user data folder in your WebView2 application
title: WebView2 Handling User Data Folder
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 05/21/2020
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, win32 apps, win32, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html, user data folder
---

# Handling the User Data Folder

> Each WebView2 instance is associated with a user data folder which stores all browser data, like cookies and network cache. 

Reviewer: Can we include an example scenario to demonstrate when developers may want to specify where user data folders are stored?

## Creation

Specify the user data folder using the **userDataFolder** parameter when creating **CoreWebView2Environment.** WebView2 stores browser data in a subfolder of the specified folder.

> When no userDataFolder is specified, WebView2 will use a default user data folder. 

Reviewer: Can we include code snippets showing the call to CoreWebView2Environment with the userDataFolder parameter? We can include one for each language/framework using tabs.

> When no userDataFolder is specified, WebView2 will use a default user data folder. 

Author: lets do this after the API has completed API Review

- For packaged Windows Store apps, the default user folder will be the **LocalFolder** sub folder of the package’s **ApplicationData folder.** 
- For legacy desktop apps, the default user data folder is the **exe path + “.WebView2”**. For legacy desktop apps, we highly recommend you specify a user data folder when creating **CoreWebView2Environment.** This folder should be in the same folder all other app data is stored. 

## Deletion

When uninstalled, the application should delete the user data folder. For packaged Windows Store apps, this is done automatically by Windows.

The app could also delete the user data folder to clean up all browsing data, to recover from possible data corruption, or to remove traces of past sessions. 

> [!NOTE]

> > Files in the user data folder can still be in use for a short while after WebView2 is closed. In this situation, wait for the associated browser process and all of its child process to exit before retry deleting of the folder. The process id of the browser process can be retrieved as **BrowserProcessId** property of the WebView2.

Reviewer: Is this done using Windows Task Manager or code? Can we add that clarification to the doc? If code, can we include a code snippet?

> > Files in the user data folder can still be in use for a short while after WebView2 is closed. In this situation, wait for the associated browser process and all of its child process to exit before retry deleting of the folder. The process id of the browser process can be retrieved as **BrowserProcessId** property of the WebView2.

Reviewer: Can we add a code snippet to show how to delete the user data folder?

## Management

All WebView2s associated with the same user data folder will run in the same browser process and share the same browsing history and cache. As a result:

> 1. When handling the **NewBrowserVersionAvailable** event to recreate WebView2 with the newer browser version, wait for the old browser process to exit and close all WebView2s sharing the same user data folder before recreating WebView2 to run from newer version. The process id of the browser process can be retrieved by **BrowserProcessId** property of the WebView2.

Reviewer: Is this done using Windows Task Manager or code? Can we add that clarification to the doc? If code, can we include a code snippet?

---
description: Learn how to manage user data folders in WebView2 applications
title: Manage user data folder in WebView2 applications.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 06/02/2020
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, win32 apps, win32, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html, user data folder
---

# Managing the User Data Folder

WebView2 applications interact with user data folders to store browser data, such as cookies, permissions, and cached resources. Each instance of a WebView2 control is associated with a user data folder. Each user data folder is unique to a user.

> ## Best Practices

Author: Why move to the bottom? My thinking was its best to read best practices first especially for architects designing the application before doing the mechanics of building the application.

> ## Best Practices

Reviewer: @pagoe-msft - when I approached the topic, I thought of learning the basics first, and then learning the more advanced content (best practices). For example, someone new to user data folders may try to build a small sample first, and then design their production application to use it. I think you bring up a good point that this content can help customers design the solution. Would changing the title to something like "Plan..." or "Design..." work?

WebView2 developers control the lifetime of the user data folder. User data folders are created automatically by WebView2. They should persist if the application relies on the user data in the application session. Otherwise they should be removed.

Here are some scenarios:

> *   If the application is repeatedly used by the same user and the web content of the application relies on the user's data, it is best if the user data folder persists.
> *   If the application does not have any repeat users, it is best to create a new user data folder for each incoming user and delete the previous user data folder.
> *   If the application has multiple repeat users, it is best to create a new user data folder for each incoming user and allow the existing user data folders to persist.

Reviewer: ```suggestion
*   If the same user uses your application repeatedly, and the web content of the application relies on the user’s data, save the user data folder. If multiple users use your application repeatedly, create a new user data folder for each new user, and save the user data folder of the previous user.
*   If your application does not have repeat users, create a new user data folder for each user, and delete the previous user data folder.
```

Your application may need to delete user data folders:

* When uninstalling your app. If you are uninstalling packaged Windows Store apps, Windows deletes user data folders automatically. 
* To clean up all browsing data history.
* To recover from data corruption.
* To remove previous session data. 

> [!NOTE]
> Files in user data folders may still be in use after the WebView2 application is closed. In this situation, wait for the browser process and all child processes to exit before deleting the folder. You may retrieve the process id of the browser process using the `BrowserProcessId` property of the WebView2.

## Share user data folders

WebView2 controls may share the same user data folders to:

* Optimize system resources by running in one browser process.
* Share browser history and cached resources. 

Consider the following when sharing user data folders: 

1. When re-creating WebView2 controls to update browser versions using `NewBrowserVersionAvailable` events, ensure browser processes exit and close WebView2 controls that share the same user data folder. To retrieve the process id of the browser process, use the `BrowserProcessId` property of the WebView2 control.

> 2. WebView2s sharing the same user data folder must use the same options for **CoreWebView2Environment**. If not, the WebView2 creation will fail with `HRESULT_FROM_WIN32(ERROR_INVALID_STATE)`. For example, if the first WebView2 is created by **CoreWebView2Environment** with a certain **AdditionalBrowserArguments**, trying to create WebView2 from **CoreWebView2Environment** with different **AdditionalBrowserArguments** will fail.

Reviewer: ```suggestion
2. WebView2 controls that share the same user data folder must use the same options for **CoreWebView2Environment**. If not, the WebView2 creation will fail with `HRESULT_FROM_WIN32(ERROR_INVALID_STATE)`. For example, if the first WebView2 is created by **CoreWebView2Environment** with certain **AdditionalBrowserArguments**, trying to create WebView2 from **CoreWebView2Environment** with different **AdditionalBrowserArguments** will fail.
```

To isolate different parts of your application or when sharing data between WebView2 controls is not needed, you may choose to use different user data folders. For example, an application may consist of two WebView2 controls, one for displaying ads and the other for displaying application content. In this scenario, developers may opt to use different user data folders for each WebView2 control. 

> > [!NOTE]

Reviewer: Please add a See Also section with some related links. Thanks.

Author: What links do you recommend?

Reviewer: @pagoe-msft - Can we link to the getting started topics?
